### PR TITLE
Aura: Ensure we are building on each relay chain fork

### PR DESCRIPTION
**Pull request by bkchr**
We only want to build one block per slot for Aura on parachains. However, we still need to build on each relay chain fork, which is using the same slot.

Closes: https://github.com/paritytech/polkadot-sdk/issues/5349

### CODE REVIEW

**Comment by burdges** [comment]
This seems fine imho.  There is techncially a DoS vector here of course, but mostly we design messaging and bridges to prefent exploiting that sort of thing.

File: cumulus/client/consensus/aura/src/collators/basic.rs [hunk 1]
```diff
@@ -215,11 +216,13 @@ where
 			//
 			// Most parachains currently run with 12 seconds slots and thus, they would try to
 			// produce multiple blocks per slot which very likely would fail on chain. Thus, we have
-			// this "hack" to only produce on block per slot.
+			// this "hack" to only produce on block per slot per relay chain fork.
```
Comment (davxy) [inline cumulus/client/consensus/aura/src/collators/basic.rs:219]:
```suggestion
			// this "hack" to only produce one block per slot per relay chain fork.
```

**Review by davxy — APPROVED** [review]

File: cumulus/client/consensus/aura/src/collators/basic.rs [hunk 2]
```diff
@@ -215,11 +216,13 @@ where
...
 			//
 			// With https://github.com/paritytech/polkadot-sdk/issues/3168 this implementation will be
 			// obsolete and also the underlying issue will be fixed.
-			if last_processed_slot >= *claim.slot() {
+			if last_processed_slot >= *claim.slot() &&
+				last_relay_chain_block > *relay_parent_header.number()
```
Comment (skunert) [inline cumulus/client/consensus/aura/src/collators/basic.rs:224]: Maybe my brain is playing me, but I think it needs to be inverted.
```suggestion
				last_relay_chain_block < *relay_parent_header.number()
```

Basically we want to skip if we are in the same slot as the previously build block and the relay parent we are building on has a higher number than the last one. If we are in the same slot but the relay parent has the same number as the last one, we are in a fork and want to build.

**Review by skunert — APPROVED** [review]

**Review by alexggh — APPROVED** [review]

**Comment by burdges** [comment]
We could imagine parachains not being able to do this, like because of forks with different ancestors, and blocks being expensive.  In particular we'd maybe lose this if we ever added some really CPU heavy storage.  What goes wrong in that situation?

**Pull request merged**